qualtran/serialization/data_types_test.py

```diff
@@ -0,0 +1,61 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import pytest
+import sympy
+from typing import Union
+from qualtran.serialization.data_types import data_type_from_proto, data_type_to_proto
+from qualtran import QDType, QBit, QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, QFxp
+
+def round_trip(data: QDType):
+    serialized_qbit = data_type_to_proto(data)
+    original_qbit = data_type_from_proto(serialized_qbit)
+    
+    assert data.num_qubits == original_qbit.num_qubits
+    if isinstance(serialized_qbit, BoundedQUInt):        
+        assert data.iteration_length == original_qbit.iteration_length
+    if isinstance(serialized_qbit, QFxp):
+        assert data.num_frac == original_qbit.num_frac
+        assert data.signed == original_qbit.signed
```

Comment:
Can we just do value equality? i.e. `assert data == original_qbit` ?

Reply:
That's true for general python classes, but in this case all `QDType` classes are implemented using `attrs.frozen` decorator which automatically overrides the `__eq__` operator to support value equality. 

```
In [3]: QFxp(3, 2) == QFxp(3, 2)
Out[3]: True
```